File: crazyflie_controllers/scripts/attitude_pid_controller.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from dataclasses import dataclass
from crazyflie_msgs.msg import AttitudeCommand
from crazyflie_msgs.msg import MotorVel
from geometry_msgs.msg import PointStamped
from sensor_msgs.msg import Imu
from tf_transformations import euler_from_quaternion

GRAVITY_COMPENSATION = 55.0
CONTROLLER_FREQ = 50
    
class PID:

    def __init__(self,kp=0,ki=0,kd=0) -> None:
        self._kp = kp
        self._ki = kd
        self._kd = kd
        self._distance_previous_error = 0
        self._count = 0

    def get_command(self, value, target):
        
        distance_error = target - value
        pid_P = self._kp * distance_error
        print(""+str(self._count) + " " +str(distance_error))
        self._count += 1
        
        pid_I = 0
        
        dist_difference = (distance_error - self._distance_previous_error) * CONTROLLER_FREQ
        pid_D = self._kd * dist_difference
        self._distance_previous_error = distance_error

        command = pid_P + pid_I + pid_D
        return command




class AttitudePID(Node):
    
    def __init__(self) -> None:
        
        super().__init__('attitude_pid_controller')

        self._x = 0
        self._y = 0
        self._z = 0
        self._roll = 0
        self._pitch = 0
        self._yaw = 0
    
        self._sub_new_position = self.create_subscription(
                AttitudeCommand ,"/crazyflie/pid/attitude_controller", self._new_command_callback,10)
        self._sub_gps = self.create_subscription(
                PointStamped, "/crazyflie/gps", self._new_gps_callback, 10 )
        self._sub_imu = self.create_subscription( Imu,"/crazyflie/imu", self._new_imu_callback, 10)
        self._pub_motor_vel = self.create_publisher(
                MotorVel, "/crazyflie/cmd_motor_vel", 10 )
        
        self._cmd_motor_timer = self.create_timer(1/CONTROLLER_FREQ, self._send_cmd_motor)
        
        self._pid_z = PID(kp=0.3,ki=0,kd=50)
        self._pid_roll = PID(kp=2,ki=0,kd=30)
        self._pid_pitch = PID(kp=2,ki=0,kd=30)
        
        self._target_z = 2
        self._target_roll = 0
        self._target_pitch = 0
        self._target_yaw = 0
        


    
    def _new_command_callback(self, command : AttitudeCommand):
        
        self._target_pitch = command.pitch
        self._target_roll = command.roll
        self._target_yaw = command.yaw
        self._target_z = command.thurst


    def _new_imu_callback(self, imu_data : Imu):

        quaternion = [
            imu_data.orientation.x,
            imu_data.orientation.y,
            imu_data.orientation.z,
            imu_data.orientation.w]
        (self._roll, self._pitch, self._yaw) = euler_from_quaternion(quaternion)
        self.get_logger().debug(f"imu data received! \
            [r: {self._roll}, p: {self._pitch}, y: {self._yaw}]")
        

    def _new_gps_callback(self, gps_data : PointStamped):
        
        self._x = gps_data.point.x
        self._y = gps_data.point.y
        self._z = gps_data.point.z
        self.get_logger().debug(f"gps data received! \
            [x: {gps_data.point.x}, y: {gps_data.point.y}, z: {gps_data.point.z}]")
        
        

             

    def _send_cmd_motor(self):

        throttle = self._pid_z.get_command(value=self._z, target=self._target_z) 
        #roll = self._pid_roll.get_command(value=self._roll, target=self._target_roll)
        #pitch = self._pid_pitch.get_command(value=self._pitch, target=self._target_pitch)

        msg = MotorVel()
        msg.m1 = GRAVITY_COMPENSATION + throttle #- roll
        msg.m2 = GRAVITY_COMPENSATION + throttle #- roll
        msg.m3 = GRAVITY_COMPENSATION + throttle #+ roll
        msg.m4 = GRAVITY_COMPENSATION + throttle #+ roll
        self._pub_motor_vel.publish(msg)
   

def main(args=None):
    rclpy.init(args=args)
    node = AttitudePID()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == "__main__":
    main()
    

    

File: crazyflie_controllers/src/attitude_pid_controller.cpp
#include "crazyflie_controllers/attitude_pid_controller.h"

AttitudePID::AttitudePID()
    : Node("attitude_pid_controller"), _input_thrust(1.0), _target_roll(0.0), _target_pitch(0.0), _target_yaw(0.0),
     _pid_roll(0, 0, 0), _pid_pitch(0.8, 0, 0), _pid_yaw(0, 0, 0)
{
        _sub_new_position = this->create_subscription<crazyflie_msgs::msg::AttitudeCommand>(
            "/crazyflie/pid/attitude_controller",
            10,
            std::bind(&AttitudePID::_newCommandCallback, this, std::placeholders::_1));

        _sub_gps = this->create_subscription<geometry_msgs::msg::PointStamped>(
            "/crazyflie/gps",
            10,
            std::bind(&AttitudePID::_newGpsCallback, this, std::placeholders::_1));

        _sub_imu = this->create_subscription<sensor_msgs::msg::Imu>(
            "crazyflie/imu",
            10,
            std::bind(&AttitudePID::_newImuCallback, this, std::placeholders::_1));

        _pub_motor_vel = this->create_publisher<crazyflie_msgs::msg::MotorVel>(
            "/crazyflie/cmd_motor_vel",
            10);

        _cmd_motor_timer = this->create_wall_timer(
            std::chrono::milliseconds(1000 / CONTROLLER_FREQ),
            std::bind(&AttitudePID::_sendCmdMotor, this));


        _sub_tuner = this->create_subscription<crazyflie_msgs::msg::PidTuner>(
            "/crazyflie/pid_tuner",
            10,
            std::bind(&AttitudePID::_newPIDTunerCallback, this, std::placeholders::_1));

        _old_time = now();
    
}

void AttitudePID::_newPIDTunerCallback(const crazyflie_msgs::msg::PidTuner::SharedPtr pid_tune_data) {
    this->_pid_pitch.update(pid_tune_data->pitch_kp, pid_tune_data->pitch_ki, pid_tune_data->pitch_kd);
    this->_pid_roll.update(pid_tune_data->roll_kp, pid_tune_data->roll_ki, pid_tune_data->roll_kd);
    this->_pid_yaw.update(pid_tune_data->yaw_kp, pid_tune_data->yaw_ki, pid_tune_data->yaw_kd);
}

void AttitudePID::_newCommandCallback(const crazyflie_msgs::msg::AttitudeCommand::SharedPtr command) {
    _target_pitch = command->pitch;
    _target_roll = command->roll;
    _target_yaw = command->yaw;
    _input_thrust = command->thurst;
    RCLCPP_DEBUG(this->get_logger(), "target updated");
}

void AttitudePID::_newImuCallback(const sensor_msgs::msg::Imu::SharedPtr imu_data) {
    tf2::Quaternion quaternion(
        imu_data->orientation.x,
        imu_data->orientation.y,
        imu_data->orientation.z,
        imu_data->orientation.w);

    tf2::Matrix3x3 mat(quaternion);
    mat.getRPY(_roll, _pitch, _yaw);

    RCLCPP_INFO(this->get_logger(), "imu data received! [r: %f, p: %f, y: %f]", _roll, _pitch, _yaw);
    RCLCPP_INFO(this->get_logger(), "p: %f", _pitch);
}

void AttitudePID::_newGpsCallback(const geometry_msgs::msg::PointStamped::SharedPtr gps_data) {
    _x = gps_data->point.x;
    _y = gps_data->point.y;
    _z = gps_data->point.z;
    RCLCPP_DEBUG(this->get_logger(), "gps data received! [x: %f, y: %f, z: %f]", _x, _y, _z);
}


void AttitudePID::_sendCmdMotor() {

    rclcpp::Duration dt = now() - _old_time; 
    
    double pid_roll     = _pid_roll.getCommand(_roll, _target_roll, dt);
    double pid_pitch    = - _pid_pitch.getCommand(_pitch, _target_pitch, dt);
    double pid_yaw      = _pid_yaw.getCommand(_yaw, _target_yaw, dt);

    auto msg = std::make_unique<crazyflie_msgs::msg::MotorVel>();
    // RCLCPP_INFO(this->get_logger(), "data received thrust %f", _pitch );
    auto mapped_thrust = ((_input_thrust - 2.5) / 7.5) / 1.5;
    msg->m1 = mapped_thrust - pid_roll - pid_pitch + pid_yaw; // GRAVITY_COMPENSATION + _input_thrust - pid_roll - pid_pitch + pid_yaw;
    msg->m2 = mapped_thrust - pid_roll + pid_pitch - pid_yaw; // GRAVITY_COMPENSATION + _input_thrust - pid_roll + pid_pitch - pid_yaw;
    msg->m3 = mapped_thrust + pid_roll + pid_pitch + pid_yaw; // GRAVITY_COMPENSATION + _input_thrust + pid_roll + pid_pitch + pid_yaw;
    msg->m4 = mapped_thrust + pid_roll - pid_pitch - pid_yaw; // GRAVITY_COMPENSATION + _input_thrust + pid_roll - pid_pitch - pid_yaw;
    _pub_motor_vel->publish(std::move(msg));

    _old_time = now();
}

int main(int argc, char** argv) 
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<AttitudePID>());
    rclcpp::shutdown();
    return 0;
}
